src/background/transactions/TransactionService.ts

```diff
@@ -41,20 +57,126 @@ export class TransactionService {
+      // the whole background script runs from scratch? If it does, it means we
+      // instantiate TransactionService anyway, and when we do, we {performPurgeCheck()} anyway...
+      // So do we need a handler at all?
+      // But if we don't add a handler, might chrome be "smart" and not run the alarm?
+      // Documentation doesn't have answers for this.
+      this.performPurgeCheck();
+    });
   }
 
   async initialize() {
     await this.transactionsStore.ready();
-    const pending = getPendingTransactions(this.transactionsStore.getState());
+    const transactions = this.transactionsStore.getState();
+    const pending = getPendingTransactions(transactions);
     this.transactionsPoller.add(pending.map(toPollingObj));
     this.addListeners();
+    if (transactions.length) {
+      this.startPurgeInterval({ leading: true });
+    }
+  }
+
+  private startPurgeInterval({ leading } = { leading: false }) {
+    if (leading) {
+      this.performPurgeCheck(); // make leading call
+    }
+    TransactionService.scheduleAlarms();
+  }
+
+  private schedulePurgeCheck = throttle(
+    () => {
+      this.performPurgeCheck();
+    },
+    FOUR_MINUTES_IN_MS,
+    { leading: false } // Invoke no sooner and no more frequent than FOUR_MINUTES
+  );
+
+  /**
+   * Purges from cache all transactions belonging to {address} in {chainId}
+   * which have a nonce less than or equals to {fromNonce}
+   */
+  private async purgeEntries({
+    address,
+    fromNonce,
+    chainId,
+  }: {
+    address: string;
+    fromNonce: number;
+    chainId: ChainId;
+  }) {
+    const transactions = await this.transactionsStore.getSavedState();
+    const candidates = transactions.filter((item) => {
+      return (
+        normalizeAddress(item.transaction.from) === normalizeAddress(address) &&
+        normalizeChainId(item.transaction.chainId) === chainId &&
+        item.transaction.nonce <= fromNonce
+      );
+    });
+    this.transactionsStore.bulkDeleteTransactionsByHash(
+      candidates.map((item) => item.hash)
+    );
+  }
+
+  /** Finds local transactions which our backend is aware of and removes them */
+  private async performPurgeCheck() {
+    const transactions = await this.transactionsStore.getSavedState();
+
+    type Address = string;
+    type Key = `${Address}:${ChainId}`;
+    const map = new Map<Key, { hash: string; timestamp: number }>();
+
+    for (const item of transactions) {
+      const chainId = normalizeChainId(item.transaction.chainId);
+      const key = `${item.transaction.from}:${chainId}` as const;
+      map.set(key, { hash: item.hash, timestamp: item.timestamp });
+    }
+
+    for (const [key, { hash, timestamp }] of map.entries()) {
+      const [address, chainIdStr] = key.split(':');
+      const chainId = chainIdStr as ChainId;
+      const network = await getNetworkByChainId(chainId);
+      if (network?.supports_actions) {
+        const knownNonce = await getLatestNonceKnownByBackend({
```

Comment:
Will we make a call for all actions one by one here?
Should we somehow exclude removed hashes from the `map`?